packages/languages/src/node/language-server-contribution.ts

```diff
@@ -48,27 +62,24 @@ export abstract class BaseLanguageServerContribution implements LanguageServerCo
         return message;
     }
 
-    protected createProcessSocketConnection(
-        outSocket: MaybePromise<net.Socket>, inSocket: MaybePromise<net.Socket>,
-        command: string, args?: string[], options?: cp.SpawnOptions
-    ): Promise<IConnection> {
+    protected async createProcessSocketConnection(outSocket: MaybePromise<net.Socket>, inSocket: MaybePromise<net.Socket>,
+        command: string, args?: string[], options?: cp.SpawnOptions): Promise<IConnection> {
+
         const process = this.spawnProcess(command, args, options);
-        return Promise.all([
-            Promise.resolve(outSocket),
-            Promise.resolve(inSocket)
-        ]).then(result => createProcessSocketConnection(process, result[0], result[1]));
+        const [outSock, inSock] = await Promise.all([outSocket, inSocket]);
+        return createProcessSocketConnection(process.process, outSock, inSock);
     }
 
     protected createProcessStreamConnection(command: string, args?: string[], options?: cp.SpawnOptions): IConnection {
         const process = this.spawnProcess(command, args, options);
-        return createProcessStreamConnection(process);
+        return createStreamConnection(process.output, process.input, () => process.kill());
     }
 
-    protected spawnProcess(command: string, args?: string[], options?: cp.SpawnOptions): cp.ChildProcess {
-        const serverProcess = cp.spawn(command, args, options);
-        serverProcess.once('error', this.onDidFailSpawnProcess.bind(this));
-        serverProcess.stderr.on('data', this.logError.bind(this));
-        return serverProcess;
+    protected spawnProcess(command: string, args?: string[], options?: cp.SpawnOptions): RawProcess {
+        const rawProcess = this.processFactory({ command, args, options });
```

Comment:
when will this process be deleted from the manager? as I read the code it seems never except `onStop`

Reply:
Removing them should be managed by the ProcessManager, since we might want to keep a certain amount of history about closed processes.

Reply:
We decided to automatically unregister processes when they are killed until we have a different way of cleaning up terminated processes in the ProcessManager.

Reply:
I don't see the process being killed in the code ?

Also I think it may be good to unregister the process after all automatically but it should be done in a on('exit' callback and registered by the processManager.

This is fine a long as the Process is not disposable.  That way even if it's unregistered references to that process still exists and will receive the onDeleted event.

Reply:
It is killed then the connection is closed:
https://github.com/theia-ide/theia/blob/3ee55262a0d2a2758e680d1d845b81e746e86b48/packages/languages/src/node/language-server-contribution.ts#L71

during forwarding, if one connection get closed another will be closed as well

Reply:
the on('exit' could also be in the Process class

Reply:
Ha ok thx my comments still apply